### PR TITLE
Refactor of aobasis

### PR DESCRIPTION
**Pull request by JensWehner**
- Removed all the recursion in basisset.cc and replaced it with for loops for better readability and added some tests
- Moved ReorderMOs from AOBasis to QMPackage
- Input format of types is now checked
- The Reorder parameters are now stored in the qmpackage and not hardcoded in aobasis.cc

### CODE REVIEW

File: src/libxtp/basisset.cc [hunk 1]
```diff
@@ -23,136 +23,100 @@ namespace votca {
...
-      lmax = 4;
-    } else if (type == "H") {
-      lmax = 5;
-    } else if (type == "I") {
-      lmax = 6;
-    } else {
-      throw std::runtime_error("FindLmax: Shelltype not known");
-    }
+
+  assert(!type.empty() && "Shelltype most be non empty!");
```
Comment (baumeier) [inline src/libxtp/basisset.cc]: “must” :)

File: src/libxtp/basisset.cc [hunk 2]
```diff
@@ -23,136 +23,100 @@ namespace votca {
...
+  } else if (t == 'I') {
+    lmax = 6;
   } else {
-    for (Index i = 0; i < Index(type.length()); ++i) {
-      std::string local_shell = std::string(type, i, 1);
-      Index test = FindLmax(local_shell);
-      if (test > lmax) {
-        lmax = test;
-      }
-    }
```
Comment (baumeier) [inline src/libxtp/basisset.cc]: What will happen if there is a SP shell or similar in the basis?

Comment (JensWehner) [reply src/libxtp/basisset.cc]: it always takes the last character of the shell. These routines now assume that a) the shell characters are sorted and continous.

Comment (baumeier) [reply src/libxtp/basisset.cc]: Makes sense.

File: src/libxtp/basisset.cc [hunk 3]
```diff
@@ -23,136 +23,100 @@ namespace votca {
...
-      std::string local_shell = std::string(type, i, 1);
-      Index test = FindLmin(local_shell);
-      if (test == 0) {
-        return 0;
-      }
-      if (test < lmin) {
-        lmin = test;
-      }
-    }
+    throw std::runtime_error("FindLmax: Shelltype not known");
```
Comment (baumeier) [inline src/libxtp/basisset.cc]: FindLmin

**Comment by JensWehner** [comment]
I am considering doing away with all the characters for shells and just have an lmin and lmax. WHat do you thing?

**Comment by JensWehner** [comment]
@felipeZ please review as well.

File: include/votca/xtp/qmpackage.h [hunk 4]
```diff
@@ -118,6 +118,23 @@ class QMPackage {
   std::vector<std::string> GetLineAndSplit(std::ifstream& input_file,
                                            const std::string separators) const;
 
+  void ReorderMOsToXTP(Eigen::MatrixXd& v, const AOBasis& basis) const;
```
Comment (felipeZ) [inline include/votca/xtp/qmpackage.h]: I still think that for people new to the code a brief comment explaning what those methods do would be very useful :)

File: src/libxtp/qmpackages/orca.h [hunk 5]
```diff
@@ -57,7 +57,29 @@ class Orca : public QMPackage {
...
 
+ protected:
+  const std::array<Index, 25>& ShellMulitplier() const final {
+    return _multipliers;
+  }
+  const std::array<Index, 25>& ShellReorder() const final { return _reorder; }
+
  private:
+  // clang-format off
+  std::array<Index,25> _multipliers={
```
Comment (felipeZ) [inline src/libxtp/qmpackages/orca.h]: Could you please explain to me what are those multipliers?

File: src/libxtp/basisset.cc [hunk 6]
```diff
@@ -174,70 +138,75 @@ std::vector<Index> NumFuncSubShell(const std::string& shell_type) {
...
-    } else {
-      throw std::runtime_error(
-          "OffsetFuncShell_cartesian shell_type not known");
-    }
+bool CheckShellType(const std::string& shelltype) {
+  if (shelltype.empty()) {
+    return false;
+  }
+  std::vector<char> allowed_shells = {'S', 'P', 'D', 'F', 'G', 'H', 'I'};
+  std::vector<char>::iterator it =
```
Comment (felipeZ) [inline src/libxtp/basisset.cc]: **auto** can help here to reduce the *Carpal tunnel syndrome* ;)

File: src/libxtp/qmpackage.cc [hunk 7]
```diff
@@ -140,5 +138,118 @@ std::string QMPackage::FindDefaultsFile() const {
   return xmlFile;
 }
 
+void QMPackage::ReorderMOs(Eigen::MatrixXd& v,
```
Comment (felipeZ) [inline src/libxtp/qmpackage.cc]: The methods from here to end of the file are some much better and cleaner! Awesome work :+1:

File: src/libxtp/basisset.cc (hunk 6 above)
Comment (JensWehner) [reply src/libxtp/basisset.cc]: :) I prefer the explicit statements a bit more to be honest. I see your point though.

File: src/libxtp/qmpackages/orca.h (hunk 5 above)
Comment (JensWehner) [reply src/libxtp/qmpackages/orca.h]: to be honest I do not really get it myself. In my mind they should not be there but @baumeier found that without them it does not work. @baumeier want to explain?

Comment (baumeier) [reply src/libxtp/qmpackages/orca.h]: Not much to explain. At some point, when we compared the coefficients coming out of orca to other codes (and our own), some entries associated to these functions showed a minus sign difference. BUT only on some elements, not the whole vector. 
Orca is using libint, right? Maybe one can look into that to see if there is a difference in definition of the affected AOs? @felipeZ don‘t you have some experience with libint?

Comment (JensWehner) [reply src/libxtp/qmpackages/orca.h]: it could actually be a different normalisation condition.

Comment (baumeier) [reply src/libxtp/qmpackages/orca.h]: Also possible, but orca is a bit too black boxy overall. I don‘t think they have these details documented.

Comment (baumeier) [reply src/libxtp/qmpackages/orca.h]: Nor do we...

Comment (JensWehner) [reply src/libxtp/qmpackages/orca.h]: it is in the code :)

Comment (JensWehner) [reply src/libxtp/qmpackages/orca.h]: ...somewhere

Comment (felipeZ) [reply src/libxtp/qmpackages/orca.h]: @baumeier  I have experience with libint2, I use it to compute multipole integrals using CP2K optimized basissets

**Pull request merged**